lib/droplet_kit/client.rb

```diff
@@ -1,30 +1,47 @@
 # frozen_string_literal: true
 
 require 'faraday'
+require 'faraday/retry'
 require 'droplet_kit/utils'
 
 module DropletKit
   class Client
     DEFAULT_OPEN_TIMEOUT = 60
     DEFAULT_TIMEOUT = 120
+    DEFAULT_RETRY_MAX = 3
+    DEFAULT_RETRY_WAIT_MIN = 0
+
     DIGITALOCEAN_API = 'https://api.digitalocean.com'
 
-    attr_reader :access_token, :api_url, :open_timeout, :timeout, :user_agent
+    attr_reader :access_token, :api_url, :open_timeout, :timeout, :user_agent, :retry_max, :retry_wait_min
 
     def initialize(options = {})
       options = DropletKit::Utils.transform_keys(options, &:to_sym)
-      @access_token = options[:access_token]
-      @api_url      = options[:api_url] || DIGITALOCEAN_API
-      @open_timeout = options[:open_timeout] || DEFAULT_OPEN_TIMEOUT
-      @timeout      = options[:timeout] || DEFAULT_TIMEOUT
-      @user_agent   = options[:user_agent]
+      @access_token   = options[:access_token]
+      @api_url        = options[:api_url] || DIGITALOCEAN_API
+      @open_timeout   = options[:open_timeout] || DEFAULT_OPEN_TIMEOUT
+      @timeout        = options[:timeout] || DEFAULT_TIMEOUT
+      @user_agent     = options[:user_agent]
+      @retry_max      = options[:retry_max] || DEFAULT_RETRY_MAX
+      @retry_wait_min = options[:retry_wait_min] || DEFAULT_RETRY_WAIT_MIN
     end
 
     def connection
       @faraday ||= Faraday.new connection_options do |req|
         req.adapter :net_http
         req.options.open_timeout = open_timeout
         req.options.timeout = timeout
+        req.request :retry, {
+          max: @retry_max,
+          interval: @retry_wait_min,
+          retry_statuses: [429],
+          # faraday-retry supports both the Retry-After and RateLimit-Reset
+          # headers, however, it favours the RateLimit-Reset one. To force it
+          # to use the Retry-After header, we override the header that it
+          # expects for the RateLimit-Reset header to something that we know
+          # we don't set.
+          rate_limit_reset_header: 'undefined'
```

Comment:
See https://github.com/lostisland/faraday-retry/blob/41b7ea27e30d99ebfed958abfa11d12b01f6b6d1/spec/faraday/retry/middleware_spec.rb#L269-L274 for confirmation, this is surprising behaviour.

Reply:
I've since spent some time testing this manually, and while you're right, I think this is still the best approach for the library.

I've restored the previous 429 status code handling if the `Retry-After` header is present, so that when the burst limit is reached, the client will wait, but if it's not, a rate limit error is raised. For this to work, this line is needed, since the `faraday-retry` middleware executes before the error handling logic. If it's not present, `faraday-retry` takes over and tries to wait until the `RateLimit-Reset` header, which can (in the absolute worst possible case) wait for an hour. The client being unresponsive for an hour makes it feel like the client is broken, so I opted to return an error in that case.

Reply:
Is there a way to override this altogether? Should this be conditional on `retry_max` being set to greater than 0?

Reply:
Update to be conditional on `retry_max` being set greater than 0.